sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -283,3 +303,166 @@ def get_lpmode_support(self):
+            speed = 25000
+        elif '10G' in ifname or 'SFI' in ifname or 'XFI' in ifname:
+            speed = 10000
+        elif '1000BASE' in ifname:
+            speed = 1000
+        return speed
+
+    def get_cmis_application_selected(self):
+        ap_code = 0
+        if not self.is_flat_memory():
+            field = "{}_{}".format(consts.STAGED_CTRL0_APSEL_FIELD, 1)
+            ap_code = self.xcvr_eeprom.read(field) >> 4
+
+        return (ap_code & 0xf)
+
+    def get_cmis_application_matched(self, host_speed, host_lanes):
+        if host_speed == 0 or host_lanes == 0:
+            return 0
+
+        ap_code = 0
+        appl_dict = ast.literal_eval(self.xcvr_eeprom.read(consts.APPLS_ADVT_FIELD))
+        for c in appl_dict.keys():
+            d = appl_dict[c]
+            if d.get('host_lane_count') != host_lanes:
+                continue
+            if self.get_host_speed(d.get('host_electrical_interface_id')) != host_speed:
+                continue
+            ap_code = c
+
+        return (ap_code & 0xf)
+
+    def set_cmis_application(self, host_speed, host_lanes):
+        if self.is_flat_memory():
+            return False
+
+        # Only applicable to CMIS v4.0 onwards
+        if self.xcvr_eeprom.read(consts.CMIS_REVISION_FIELD) < 0x40:
+            return False
+
+        # Check for application selection updates
+        ap_new = self.get_cmis_application_matched(host_speed, host_lanes)
+        ap_new = ap_new if ap_new > 0 else 1
+        ap_old = self.get_cmis_application_selected()
+        ap_old = ap_old if ap_old > 0 else 1
+        if ap_old == ap_new:
+            if ap_new == 1:
+                pst_ok = [0, CMIS_DATAPATH_STATE_ACTIVATED]
+                cst_ok = [0, CMIS_CONFIG_STATE_ACCEPTED]
+            else:
+                pst_ok = [CMIS_DATAPATH_STATE_ACTIVATED]
+                cst_ok = [CMIS_CONFIG_STATE_ACCEPTED]
+            pst = self.xcvr_eeprom.read(consts.DATAPATH_STATE_FIELD)
+            cst = self.xcvr_eeprom.read(consts.CONFIG_ERROR_FIELD)
+            if pst in pst_ok and cst in cst_ok:
+                # No update to the application selection, skipping...
+                return True
+
+        ready = False
+        for retries in range(2):
+            if ready:
+                break
+
+            # Reset the module
+            self.reset()
+
+            # Deinitialize DataPath
+            self.xcvr_eeprom.write(consts.DATAPATH_DEINIT_FIELD, 0xff)
+
+            # Turn off Low-Power Mode
+            self.xcvr_eeprom.write(consts.MODULE_LEVEL_CTRL_LPSW_FIELD, 0)
+            self.xcvr_eeprom.write(consts.MODULE_LEVEL_CTRL_LPHW_FIELD, 0)
+
+            for lane in range(1, 9):
+                # Update the ApCode
+                val = ap_new << 4
+                # Update the 1st lane of each datapath
+                if ap_new > 1:
+                    val |= (int((lane - 1) / host_lanes) * host_lanes) << 1
+                # Toggle the BIT0(i.e. application-defined)
```

Comment:
Why is this needed?

Reply:
Some of the CMIS4 optics are using faulty firmware design, the BIT0 need to be set to activate the selected application, while some are expecting BIT0 to be cleared.